File: solutions/elixir/140-longest-increasing-subsequence.ex
# Problem 140: Longest Increasing Subsequence (LeetCode 300)
# Difficulty: Med
# Language: Elixir
# 
defmodule Solution do
  def length_of_lis(nums) do
    Enum.reduce(nums, [], fn num, tails ->
      pos = bisect(tails, num, 0, length(tails))
      if pos == length(tails) do
        tails ++ [num]
      else
        List.replace_at(tails, pos, num)
      end
    end) |> length()
  end

  defp bisect(_, _, lo, hi) when lo >= hi, do: lo
  defp bisect(tails, target, lo, hi) do
    mid = lo + div(hi - lo, 2)
    if Enum.at(tails, mid) < target,
      do: bisect(tails, target, mid + 1, hi),
      else: bisect(tails, target, lo, mid)
  end
end

File: solutions/scala/140-longest-increasing-subsequence.scala
/**
 * Problem 140: Longest Increasing Subsequence (LeetCode 300)
 * Difficulty: Med
 * Language: Scala
 */
def lengthOfLIS(nums: Array[Int]): Int = {
  import java.util
  nums.foldLeft(Vector.empty[Int]) { (tails, num) =>
    val pos = {
      var (lo, hi) = (0, tails.length)
      while (lo < hi) { val mid = (lo + hi) / 2; if (tails(mid) < num) lo = mid + 1 else hi = mid }
      lo
    }
    if (pos == tails.length) tails :+ num
    else tails.updated(pos, num)
  }.length
}
